producer/buffer/buffer.go

```diff
@@ -270,6 +294,64 @@ func (b *buffer) cleanupBatchWithLock(
 	return next, removed
 }
 
+func (b *buffer) dropEarliestUntilClose() {
+	ticker := time.NewTicker(b.opts.DropEarliestInterval())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			select {
+			case <-b.dropEarliestCh:
+			default:
+				continue
+			}
+			b.dropEarliestUntilTarget(b.maxBufferSize)
+		case <-b.doneCh:
+			return
+		}
+	}
+}
+
+func (b *buffer) dropEarliestUntilTarget(targetSize uint64) {
+	shouldContinue := true
+	for shouldContinue {
+		b.listLock.Lock()
```

Comment:
Any reason you need to lock and unlock multiple times?

Reply:
to release the lock and allow other threads (most likely new writes) to proceed